Add an extraction of commit messages

bockthom opened the pull request:
Extract the commit messages from the database together with the corresponding commit hash into a separate file.
The commit-message extraction is implemented for both project level and range level.
To use the commit-message extraction, an additional command-line option "-m" or "--commit-messages" is added.

In addition, don't print log statements regarding range-level extractions when no range-level extractions are performed.

bockthom committed:
Add an extraction of the commit messages

Extract the commit messages together with the corresponding commit hash.

The commit-message extraction is implemented for both project level and range level.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Apply function "remove_problematic_charaters" to commit-message extraction

As their occur also encoding issues in the extracted commit messages,
also call the function "remove_problematic_characters" for
commit-message extraction.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Rename function "remove_problematic_characters"

As the function does not only remove problematic characters, but also
fixes the character encoding in general, rename the function to
"fix_characters_in_string".

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Add command-line parameter for commit-message extraction

Similar to the implementation extraction, also add a command-line
parameter for enabling the commit-message extraction.

From now on, use the parameter -m oder --commit-messages to extract the
commit messages. Otherwise, they are not extracted any more.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Fix log output when not performing range-level extractions

When range-level extractions are *not* enabled (the default), the
statement "Extracting data for version '{hash}'" still appeared in the
log output. As no range data is extracted, also the log statement should
not be printed. This commit fixes that and also the list of revisions is
only checked when the range-level extractions are enabled.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

clhunsen commented on codeface_extraction/codeface_extraction.py:
```diff
@@ -50,34 +50,39 @@ def run_extraction(conf, resdir, extract_impl, extract_on_range_level):
+    # run range-level extractions (only if explicitely enabled)
+    if extract_on_range_level:
+
+        # check if list of revisions in database is the same as in the config file
```
Do we still need this check (i.e., Lines 64–75)? I added it in a time when we have not introduced the selection processes yet and built networks on release ranges defined/extracted by Codeface – it kept me from overwriting old extractions for casestudies where I changed the configuration file, but forgot to run the commit analysis in Codeface. As we do not construct networks on Codeface ranges most of the time, we may want to remove this check.

bockthom replied:
Nonetheless, I would like to keep the check as it does not hurt and if someone eventually wants to use the Codeface ranges (for whatever reason), it will be a useful check.

clhunsen replied:
Okay for me. :wink:

clhunsen commented on codeface_extraction/extractions.py:
```diff
@@ -35,14 +35,21 @@
-        if (extract_impl or
-                (str(cls) != "<class 'codeface_extraction.extractions.FunctionImplementationExtraction'>" and
-                 str(cls) != "<class 'codeface_extraction.extractions.FunctionImplementationRangeExtraction'>")):
+        if ((extract_impl and
```
Unfortunately, this condition is quite elongated and confusing now. Can we do something like this instead:
```Python
# list of extractions to skip
extractions_to_skip = []
if not extract_impl:
    extractions_to_skip += ["<class 'codeface_extraction.extractions.CommitMessageExtraction'>"]
# add extractions to skip also for other flags

# collect all extractions and instantiate objects
for cls in Extraction.__subclasses__():
    __extractions.append(cls(dbm, conf, resdir, csv_writer))
# remove all extractions that are listed in extractions_to_skip
__extractions = filter(lambda x: str(x) not in extractions_to_skip, __extractions)
```

Maybe, we can also find a better way than the call to `str()` on the way.

------------------
Disclaimer: The code is untested as I do not have a proper Python installation on this computer.

bockthom replied:
I also dislike the elongated if-statement, but it was a simple and easy way to introduce configuration parameter of the additional extraction. Nevertheless, your suggestion looks good to me (except for mixing implementation extraction and commit message extraction). I will try out your code suggestion tomorrow.

bockthom replied:
Just one comment: The last line of your suggestion does not not work as we do not operate on classes any more, but on already instantiated classes, that is, objects. Nonetheless, I have just removed those line and added a simple if-statement before the `__extractions.append` statement. Please see my new commit.

ecklbarb replied:
Thank you for this nice solution, @bockthom and @clhunsen.

clhunsen reviewed (changes requested):
Thanks for your additions and changes, @bockthom! 

In general, the changes look very good, unfortunately, I have longer comments on two issues (along with suggestions on how to solve them). Feel free to discuss. Also, if you want, I can provide a patch when I am back in office tomorrow.

bockthom committed:
Improve elongated if-statement for filtering extractions

Remove the elongated if-statement which checks which extractions are
enabled and which not. Instead, construct a list of extractions to skip
and just check whether an extraction is in the list or not.

Props to @clhunsen for pointing this out and providing a suggestion how
to improve this.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>
Signed-off-by: Claus Hunsen <hunsen@fim.uni-passau.de>
